### ISSUE TITLE
Port `read-char` located in `keyboard.c`.

### ISSUE THREAD

**Issue by harryfei** [Wilfred/remacs#924]
This task is for guys who want some challenge. Kill the monster, Hero!.

**Comment by nickdrozd** [comment]
This one isn't an exposed Lisp function, so it should be `read_char` instead of `read-char`. Actually in some sense it doesn't needed to be "ported" at all. It's strictly an internal function, so it could be reimplemented in a completely different way. Not that I would be able to do that of course; I can't even understand the definition of the `sys_jmp_buf` type it uses:

```c
/* Save and restore the instruction and environment pointers,
   without affecting the signal mask.  */

#ifdef HAVE__SETJMP
typedef jmp_buf sys_jmp_buf;
# define sys_setjmp(j) _setjmp (j)
# define sys_longjmp(j, v) _longjmp (j, v)
#elif defined HAVE_SIGSETJMP
typedef sigjmp_buf sys_jmp_buf;
# define sys_setjmp(j) sigsetjmp (j, 0)
# define sys_longjmp(j, v) siglongjmp (j, v)
#else
/* A platform that uses neither _longjmp nor siglongjmp; assume
   longjmp does not affect the sigmask.  */
typedef jmp_buf sys_jmp_buf;
# define sys_setjmp(j) setjmp (j)
# define sys_longjmp(j, v) longjmp (j, v)
#endif
```

**Comment by db48x** [comment]
These macro definitions are used for accommodating platforms where the names of things vary a bit.

I would say that porting anything that uses setjmp/longjmp is going to be quite an interesting task. longjmp is basically a computed goto. You should read the man page for them. The ideal port will refactor the code so that setjmp/longjmp are never necessary.

**Comment by brotzeit** [comment]
`insert-file-contents` is also interesting. I wonder what's the reason for this function to be THAT long.

**Comment by shaleh** [comment]
Yeah @brotzeit, that is screaming to be rip into sub functions at a minimum.
